man/lindex.man
.\" 3.2
.\"  /*******************************************************************\
.\"   * Copyright (C) 1995 Lars J Aas <larsa@colargol.stud.idb.hist.no> *
.\"   * based on documentation by C J Needham <col@imdb.com> 1995,      *
.\"   * permission is granted by the authors to freely distribute       *
.\"   *                        providing no fee of any kind is charged. *
.\"  \*******************************************************************/
.\"
.TH LINDEX 1 "RELEASEDATE" " " "The Internet Movie Database vVERSIONNUMBER"
.SH NAME
lindex \- generate lists of names or lists of titles.
.SH SYNOPSYS
.B
lindex [options]
.SH DESCRIPTION
This program can be used to either to generate lists of names that appear
on filmography lists or to generate lists of titles from the movies list.
It's uses include finding exact spellings of names or titles on the lists;
finding names/titles containing a particular substring; and finding titles
released between a range of years or in a particular year.
.SH GENERAL OPTIONS
Note that if you do not specify any database options, the default is to search
all the lists in name mode.
.SS NAME MODE
.TP
.B \-acr
actors database
.TP
.B \-acs
actresses database
.TP
.B \-dir
directors database
.TP
.B \-write
writers database
.TP
.B \-comp
composers database
.TP
.B \-cine
cinematographers database
.TP
.B \-edit
editors database
.TP
.B \-prodes
production designers database
.TP
.B \-costdes
costume designers database
.TP
.B \-misc
miscellaneous database
.TP
.B \-prdcr
producers database
.SS TITLE MODE
.TP
.B \-title
This will toggle title\-mode on.
.TP
.B \-yr
Adds the year of release and sorts in chronological order.
.TP
.B \-yru
Adds the year of release, but leaves in alphabetical order.
.TP
.B \-yrfrom <yr>
Limit search to titles released during or after
.BR <yr> .
.TP
.B \-yrto <yr>
Limit search to titles released before
.BR <yr> .
.TP
.B \-yreq <yr>
Limit search to titles released in
.BR <yr> .
.TP
.B \-mrr
Add ratings report data.
.TP
.B \-smrr
Add ratings report data and sort by average ratings.
.TP
.B \-vmrr
Add ratings report data and sort by number of votes.
.TP
.B \-match
Limit search to movies with entry in ratings database.
.TP
.B \-veq <votes>
Limit search to movies with
.BR <votes> votes.
.TP
.B \-vmin <votes>
Limit search to movies with a minimum of
.BR <votes> votes.
.TP
.B \-vmax <votes>
Limit search to movies with a maximum of
.BR <votes> votes.
.TP
.B \-m
Limit search to movies only, i.e. no TV-series.
.TP
.B \-genre <s>
Limit search to titles with a genre equal to
.BR <s> .
.TP
.B \-time <s>
Limit search to titles with time equal to
.BR <s> .
.TP
.B \-prodco <s>
Limit search to titles with a production company equal to
.BR <s> .
.TP
.B \-cert <s>
Limit search to titles with a certificate equal to
.BR <s> .
.TP
.B \-mix <s>
Limit search to titles with a sound mix equal to
.BR <s> .
.TP
.B \-cntry <s>
Limit search to titles with a country equal to
.BR <s> .
.TP
.B \-rel <s>
Limit search to titles with a releasedate equal to
.BR <s> .
.TP
.B \-loc <s>
Limit search to titles with a location equal to
.BR <s> .
.TP
.B \-tech
Limit search to titles with technical info equal to
.BR <s> .
.SH MISCELLANEOUS OPTIONS
.TP
.B \-s <substring>
Restrict search to names/titles containing
.BR <substring> .
.TP
.B \-i
Become case sensitive with
.BR \-s .
.SH EXAMPLES
.TP
Generate list of all names on the cast lists:
.B
lindex \-acr \-acs \-name
.TP
Generate list of actors containing the substring "Michael":
.B
lindex \-acr \-s "Michael" \-name
.TP
Generate list of all movie/tv titles containing "Alien":
.B
lindex \-title \-s "Alien"
.TP
List all titles in database released in 1945:
.B
lindex \-title \-yr \-yreq 1945
.TP
Sort all movies released in the 1990s by average rating:
.B
lindex \-title \-yr \-yrfrom 1990 \-yrto 1996 \-smrr
.TP
List film noirs made for RKO
.B
lindex \-title \-genre Film\-Noir \-prodco RKO
.TP
List all color MGM movies released up to 1940
.B
lindex \-title \-prodco MGM \-color Color \-yrto 1940


docs/ChangeLog
-------Version 3.4 Released---------------------

Sun Mar 29 22:40:37 BST 1998

 * biographies list offset fields increased to 4 bytes following overflow
 * various constants updated

-------Version 3.3c Released---------------------

Sat Apr 26 06:51:19 BST 1997

 * added support for crew completion list
 * added support for distributors list
 * adapted mkdbs for more robust processing of filmography lists
 * added support for new business list tag

-------Version 3.3b Released---------------------

Sat Mar 15 08:09:50 GMT 1997

 * added support for ISO-Latin-1

-------Version 3.3 Released---------------------

Wed Oct 16 22:28:54 BST 1996

 * added support for AKA attributes
 * added support for writers ordering
 * moved misc section to after case in title output
 * added support for business list
 * added support for laserdisc list
 * fixed tagline bug
 * diplays URL of web page for name/title

-------Version 3.2h Released---------------------

Tue Oct  8 14:24:21 BST 1996

 * added support for special effects companies list
 * updated various system constants
 * added new movie link types 'edited into' & 'edited from'
 * added extra checks to avoid title/attribute key file corruption in mkdbs.c

-------Version 3.2f Released---------------------

Mon Jun 24 22:04:46 GMT 1996

 * added support for language list

-------Version 3.2e Released---------------------

Sat Mar 16 08:09:39 GMT 1996

 * added support for alternate versions list
 * added bio keys IT & AT

-------Version 3.2d Released---------------------

Tue Aug 15 23:23:47 GMT 1995

 * switched TitleID to unsigned long to handle >65K titles
 * fixed bug in display of literature and genre data
 * added bio key WN

-------Version 3.2c Released---------------------

Sun May 14 23:00:14 GMT 1995

 * raised MAXFILMOGRAPHIES constant to handle larger no. of entries

-------Version 3.2 Released---------------------

Sun Jan 29 22:40:08 GMT 1995

 * added full support for title searching without year/TV/V info

Tue Jan  3 23:00:45 GMT 1995

 * replaced 'biodefs' file
 * fixed bug when looking for common titles with people using -name

Mon Jan  2 21:52:08 GMT 1995

 * log file support
 * better separation of search/display code

Wed Dec 14 21:43:57 GMT 1994

 * fixed bug in 'title' for miscellaneous section
 * added support for the movie links list

Tue Dec 13 22:06:44 GMT 1994
 
 * noted when cast in credits output
 * fixed name ordering problem for names with (I)/(II) marker
 * display Actor/Actress in list -name where appropriate

Wed Dec  7 21:19:19 GMT 1994

 * added support for tag lines & complete cast list
 * put spaces around "." in cast results
 * standardized number of dashes printed out in 'title' / 'list'

Mon Dec  5 21:19:42 GMT 1994

 * multi-country productions displayed on same line

Wed Nov 30 18:14:16 GMT 1994

 * switched to new IMDB license
 * updated 'lfetch' to get new lists
 * fixed "No plot summaries message"
 * simplified MRR processing in mkdb
 * added support for locations & release date lists
 * switched 'Reviews' marker for clarity
 * added support for technical & literature lists

Mon Oct 31 21:42:03 GMT 1994
 
-------Version 3.1 Released---------------------

Mon Oct 17 19:04:26 GMT 1994
 
 * 3.1 code frozen
 * minor bug fixes with lindex

Sun Oct  9 18:37:14 GMT 1994

 * added support for countries list

Tue Oct  4 21:27:53 GMT 1994

 * added support for cast ordering info

Sun Oct  2 23:02:59 GMT 1994

 * added support for color info and sound mix lists
 * added constraint options to 'list'
 * standardized list options

Wed Sep 28 22:22:56 GMT 1994

 * re-ordered title info printing plus separated genres list
 * print genres on one line

Mon Sep 26 21:53:47 GMT 1994

 * added support for production companies list

Sun Sep 25 22:02:07 GMT 1994

 * added constraint options to 'lindex'

Tue Sep  6 20:09:52 GMT 1994

 * added support for certificates & genres list
 * fixed mkballot bug with -ar option

Tue Aug 23 20:08:04 GMT 1994

 * added support for miscellaneous filmographies list

Tue May 24 19:13:47 GMT 1994

 * moviedb-3.0 completed

Thu May 19 22:35:12 GMT 1994

 * switched to merged character names scheme

Tue May 17 21:09:45 GMT 1994

 * new 'mkdb' eliminating need for large array to hold name keys
 * removed duplicate aka titles in 'title -s -aka' queries

Wed May 11 22:11:59 GMT 1994

 * made upperCase a macro for increased speed
 * switched to Phil's new Makefiles

Wed May  4 17:46:08 GMT 1994

 * mapped case of titles/names to match key files
 * added support for running times list
 * lguide error messages

Tue May  3 21:38:30 GMT 1994
 
 * added support for Phil's imoviedb package
 * expanded compression support

Thu Apr 28 19:51:52 GMT 1994

 * finished rewrite to use new database representation

-------Version 2.9 Released---------------------

Wed Mar 16 21:39:40 GMT 1994

 * added support for producers list
 * added support for soundtracks list
 * added support for generic aka lists

Tue Mar 15 23:47:46 GMT 1994

 * replaced -crazy and -triv options with -trivia (also includes goofs & quotes)
 * added support for quotes list
 * added support for goofs list
 * handled -full error messages correctly
 * correct conditions for printing of missing plot summaries message
 * fixed lindex to search all lists when no list options given
 * -mrr no longer implies -m

Mon Mar 14 21:48:21 GMT 1994

 * new version of check-for-program
 * passed $MAKE variable to make subprocesses
 * lfetch modified to match full filenames
 * added 'keepchars' option for CLEANLISTS

-------Version 2.8 Released---------------------

Wed Dec  1 21:58:21 GMT 1993

 * man pages added to package
 * LaTeX documentation added to package

Sat Nov 27 20:53:40 GMT 1993

 * 2.8 code finished!

Thu Nov 25 18:17:00 GMT 1993

 * ported lseen to C

Tue Nov 23 19:16:48 GMT 1993

 * added gendb -votes option

Mon Nov 22 23:54:58 GMT 1993


src/lindex.c
/*============================================================================
 *
 *  Program: lindex.c 
 *
 *  Version: 3.22
 *
 *  Purpose: index list databases
 *
 *  Author:  C J Needham <col@imdb.com>     
 *
 *  Copyright (c) 1996-2003 The Internet Movie Database Inc.
 *
 *  Permission is granted by the copyright holder to distribute this program
 *  is source form only, providing this notice remains intact, and no fee
 *  of any kind is charged. This software is provided as is and there are no 
 *  warranties, expressed or implied.
 * 
 *  Options:
 *           -acr  index actors database
 *           -acs  index actress database
 *           -dir  index directors database
 *         -write  index writers database
 *          -comp  index composers database
 *          -cine  index cinematographers database
 *          -edit  index editors database
 *        -prodes  index production designers database
 *       -costdes  index costume designers database
 *         -prdcr  index producers database
 *          -misc  index miscellaneous database
 *
 *         -title  index titles database
 *
 *           -mrr  add movie ratings report info to results
 *          -smrr  add movie ratings report info & sort by ratings
 *          -vmrr  add movie ratings report info & sort by votes
 *         -match  only print title if mrr data exists
 *
 *      -vmin <n>  limit title search to movies with n+ votes
 *      -vmax <n>  limit title search to movies with at most n votes
 *       -veq <n>  limit title search to movies with n votes
 *
 *           -yru  add year of release info
 *            -yr  add year of release info & sort chronologically
 *      -yreq <y>  limit title search to movies released in year <y>
 *    -yrfrom <y>  limit title search to movies released from <y> onwards
 *      -yrto <y>  limit title search to movies released up to <y>
 *
 *   -genre <str>  limit title search by genre
 *   -keyword <str> limit title search by keyword
 *    -cert <str>  limit title search by certificate
 *    -time <str>  limit title search by running time
 *  -prodco <str>  limit title search by production company
 *   -color <str>  limit title search by color information
 *     -mix <str>  limit title search by sound mix
 *   -cntry <str>  limit title search by country of production
 *    -lang <str>  limit title search by language
 *   -sfxco <str>  limit title search by special effects company
 *    -dist <str>  limit title search by distributor
 *
 *             -m  movies only, ignore TV-series
 *             -i  case sensistive search
 *       -s <str>  match substrings
 *
 *============================================================================
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "log.h"
#include "moviedb.h"
#include "dbutils.h"
#include "ratings.h"
#include "titleinfo.h"

#define LINDEX_USAGE1 "usage: lindex [-acr -acs -dir -write -comp -cine -edit -prodes -costdes -prdcr"
#define LINDEX_USAGE2 "               -misc] [-title] [-mrr|-smrr|-vmrr] [-match] [-s <s>] [-m -i]" 
#define LINDEX_USAGE3 "              [-yr|-yru] [-yreq <year>|-yrfrom <year> -yrto <year>] [-mix <s>]"
#define LINDEX_USAGE4 "              [-genre <s>] [-keyword <s>] [-time <s>] [-prodco <s>] [-cert <s>]"
#define LINDEX_USAGE5 "              [-cntry <s>] [-color <s>] [-lang <s>]"
#define LINDEX_USAGE6 "              [-veq <votes>|-vmin <votes>|-vmax <votes>]"

int getNextLindexName ( FILE *indexFp, FILE *keyFp, struct lindexRec *rec, struct searchConstraints *constraints )
{
  char line [ MXLINELEN ] ;
  int stopFlag = FALSE ;
  NameID nameKey ;
  long offset ;

  while ( !stopFlag )
  {
    nameKey = getName ( rec -> fp ) ;
    if ( feof ( rec -> fp ) )
      break ;
    (void) getFullOffset ( rec -> fp ) ;
    (void) fseek ( indexFp, 4 * nameKey, SEEK_SET ) ;
    offset = getFullOffset ( indexFp ) ;
    (void) fseek ( keyFp, offset, SEEK_SET ) ;
    (void) fgets ( line, MXLINELEN, keyFp ) ;
    stripSep ( line ) ;
    stopFlag = TRUE ;
    if ( constraints -> subString [ 0 ] )
      if ( ! constraints -> caseSen )
      {
        if ( caseStrStr ( line, constraints -> subString ) == NULL )
  	  stopFlag = FALSE ;
      }
      else
        if ( strstr ( line, constraints -> subString ) == NULL )
  	  stopFlag = FALSE ;
  }

  if ( stopFlag )
  {
    (void) strcpy ( rec -> name, line ) ;
    return ( TRUE ) ;
  }
  else
  {
    (void) fclose ( rec -> fp ) ;
    rec -> fp = NULL ;
    return ( FALSE ) ;
  }
}


void lindexFilmographyDatabases ( int dbCount, struct lindexRec recs [ ], struct searchConstraints *constraints )
{
  FILE *indexFp, *keyFp ;
  int i, j, activeCount, eqFlag = FALSE ;
  int active [ NO_OF_FILMOGRAPHY_LISTS ] ;
  char fn [ MAXPATHLEN ] ;

  indexFp = openFile ( NAMEIDX ) ;
  keyFp = openFile ( NAMEKEY ) ;

  for ( i = 0 ; i < dbCount ; i++ )
  {
    (void) constructFilename ( fn, filmographyDefs [ recs [ i ] . listId ] . stem, NDXEXT ) ;
    recs [ i ] . fp = openFile ( fn ) ;
  }

  if ( dbCount == 1 )
    while ( getNextLindexName ( indexFp, keyFp, &recs [ 0 ], constraints ) )
      (void) printf ( "%s\n" , recs [ 0 ] . name ) ;
  else
  {
    for ( activeCount = 0 , i = 0 ; i < dbCount ; i++ )
      if ( ( active [ i ] = getNextLindexName ( indexFp, keyFp, &recs [ i ], constraints ) ) )
        activeCount++ ;

    while ( activeCount > 0 )
    {
      for ( i = 0 ; ! active [ i ] && i < dbCount ; i++ ) ;

      for ( j = 1 ; j < dbCount ; j++ )
        if ( ( active [ j ] ) && ( eqFlag = strcmp ( recs [ i ] . name, recs [ j ] . name ) ) > 0 )
	   i = j ;
	 else if ( active [ j ] && eqFlag == 0 && i !=j )
	 {
           if ( ( active [ j ] = getNextLindexName ( indexFp, keyFp, &recs [ j ], constraints ) ) == FALSE )
	     activeCount-- ;
	   else
	     j-- ;
	 }
       (void) printf ( "%s\n" , recs [ i ] . name ) ;
       if ( ( active [ i ] = getNextLindexName ( indexFp, keyFp, &recs [ i ], constraints ) ) == FALSE )
	 activeCount-- ;
     }
   }
}


int getNextLindexTitle ( FILE *dbFp, FILE *indexFp, FILE *keyFp, struct searchConstraints *constraints, char *title, TitleID *titleKey, int *yearVal )
{
  int stopFlag = FALSE ;
  long offset ;

  while ( !stopFlag )
  {
    *titleKey = getTitle ( dbFp ) ;
    if ( feof ( dbFp ) )
      break ;
    *yearVal = getInt ( dbFp ) ;
    (void) getInt ( dbFp ) ;
    (void) getAttr ( dbFp ) ;
    (void) fseek ( indexFp, *titleKey * 4, SEEK_SET ) ;
    offset = getFullOffset ( indexFp ) ;
    (void) fseek ( keyFp, offset, SEEK_SET ) ;
    (void) fgets ( title, MXLINELEN, keyFp ) ;
    stripSep ( title ) ;
    stopFlag = TRUE ;
    if ( *yearVal < constraints -> yearFrom || *yearVal > constraints -> yearTo )
      stopFlag = FALSE ;
    if ( constraints -> moviesOnly && *title == '"' )
      stopFlag = FALSE ;
    else if ( constraints -> subString [ 0 ] )
      if ( ! constraints -> caseSen )
      {
        if ( caseStrStr ( title, constraints -> subString ) == NULL )
  	  stopFlag = FALSE ;
      }
      else
        if ( strstr ( title, constraints -> subString ) == NULL )
  	  stopFlag = FALSE ;
  }
  return ( stopFlag ) ;
}


int yearLindexSort ( struct lindexTitleRec *r1, struct lindexTitleRec *r2 )
{
  if ( r1 -> year != r2 -> year )
    return ( r1 -> year - r2 -> year ) ;
  else
    return ( caseCompare ( r1 -> title, r2 -> title ) ) ;
}


int smrrLindexSort ( struct lindexTitleRec *r1, struct lindexTitleRec *r2 )
{
  float result ;

  result =  r1 -> mrr . rating - r2 -> mrr . rating ;
  if ( result == 0 )
    return ( vmrrLindexSort ( r1, r2 ) ) ;
  else if ( result > 0 )
    return ( -1 ) ;
  else
    return ( 1 ) ;
}


int vmrrLindexSort ( struct lindexTitleRec *r1, struct lindexTitleRec *r2 )
{
  if ( r1 -> mrr . votes != r2 -> mrr . votes )
    return ( r2 -> mrr . votes - r1 -> mrr . votes ) ;
  else
    return ( caseCompare ( r1 -> title, r2 -> title ) ) ;
}



void lindexTitles ( int yropt, int mrropt, struct searchConstraints *constraints ) 
{
  FILE  *dbFp, *indexFp, *keyFp, *mrrFp = NULL ;
  FILE  *titleInfoDb [ NO_OF_TITLE_INFO_LISTS ] ;
  FILE  *titleInfoIdx [ NO_OF_TITLE_INFO_LISTS ] ;
  char  title [ MXLINELEN ] ;
  int   i, count = 0, quitLoop ;
  struct mrrRec mrr ;
  struct lindexTitleRec *array ;
  int year ;
  TitleID titleKey, mrrKey, titleInfoKey [ NO_OF_TITLE_INFO_LISTS ] ;
  long titleInfoOffset [ NO_OF_TITLE_INFO_LISTS ] ;

  for ( i = 0 ; i < NO_OF_TITLE_INFO_LISTS ; i++ )
  {
    if ( constraints -> titleInfoString [ i ] [ 0 ] )
    {
      titleInfoDb [ i ] = openFile ( titleInfoDefs [ i ] . db ) ;
      titleInfoIdx [ i ] = openFile ( titleInfoDefs [ i ] . index ) ;
      readTitleInfoIndex ( titleInfoIdx [ i ], &titleInfoKey [ i ], &titleInfoOffset [ i ] ) ;
    }
    else
    {
      titleInfoDb [ i ] = NULL ;
      titleInfoIdx [ i ] = NULL ;
    }
  }

  dbFp = openFile ( MOVIEDB ) ;
  indexFp = openFile ( TITLEIDX ) ;
  keyFp = openFile ( TITLEKEY ) ;

  if ( mrropt != NONE )
  {
    mrrFp = openFile ( MRRDB ) ;
    (void) readMrrRec ( mrrFp, &mrrKey, &mrr ) ;
  }

  array = malloc ( sizeof (struct lindexTitleRec) * MAXTITLES ) ;

  while ( getNextLindexTitle ( dbFp, indexFp, keyFp, constraints, title, &titleKey, &year ) )
  {
    for ( quitLoop = FALSE, i = 0 ; !quitLoop && i < NO_OF_TITLE_INFO_LISTS ; i++ )
    {
      if ( constraints -> titleInfoString [ i ] [ 0 ] )
      {
        while ( titleInfoKey [ i ] < titleKey && readTitleInfoIndex ( titleInfoIdx [ i ], &titleInfoKey [ i ], &titleInfoOffset [ i ] ) ) ;
        if ( titleInfoKey [ i ] != titleKey )
          quitLoop = TRUE ;
        else
        {
          if ( checkConstraints ( titleInfoDb [ i ], titleInfoOffset [ i ], titleKey , constraints -> titleInfoString [ i ] ) == FALSE )
            quitLoop = TRUE ;
        }
      }
    }
    if ( quitLoop )
      continue ;
    if ( mrropt != NONE )
    {
      while ( mrrKey < titleKey && readMrrRec ( mrrFp, &mrrKey, &mrr ) ) ;
      if ( mrr . votes < constraints -> voteMin ) 
        continue ;
      if ( mrr . votes > constraints -> voteMax ) 
        continue ;
    }
    if ( yropt == NONE && mrropt == NONE )
      (void) printf ( "%s\n", title ) ;
    else if ( yropt == YRU && mrropt == NONE )
    {
      displayTitleYear ( title, year ) ;
      (void) printf ( "\n" ) ;
    }
    else if ( yropt == NONE && mrropt == MRR )
    {
      if ( mrrKey != titleKey )
      {
        if ( !constraints -> mrrMatch )
	  (void) printf ( "                                %s\n", title ) ;
      }
      else
        (void) printf ( "      %s%8d   %2.1f  %s\n", mrr . distribution, mrr . votes, mrr . rating, title ) ;
    }
    else if ( yropt == YRU && mrropt == MRR )
    {
      if ( mrrKey != titleKey )
      {
        if ( !constraints -> mrrMatch )
        {
	  (void) printf ( "                                " ) ;
          displayTitleYear ( title, year ) ;
          (void) printf ( "\n" ) ;
        }
      }
      else
      {
        (void) printf ( "      %s%8d   %2.1f  ", mrr . distribution, mrr . votes, mrr . rating ) ;
        displayTitleYear ( title, year ) ;
        (void) printf ( "\n" ) ;
      }
    }
    else 
    {
      array [ count ] . title = duplicateString ( title ) ;
      if ( mrrKey == titleKey )
        array [ count ] . mrr = mrr ;
      else
      {
        array [ count ] . mrr . votes = 0 ;
        array [ count ] . mrr . rating = 0 ;
      }
      array [ count ] . year = year ;
      count++ ;
    }
  }

  if ( yropt == YR || mrropt == SMRR || mrropt == VMRR )
  {
    if ( yropt == YR )
      qsort ( (void*) array, (size_t) count, sizeof ( struct lindexTitleRec ), (int (*) (const void*, const void*)) yearLindexSort ) ;
    else if ( mrropt == SMRR )
      qsort ( (void*) array, (size_t) count, sizeof ( struct lindexTitleRec ), (int (*) (const void*, const void*)) smrrLindexSort ) ;
    else
      qsort ( (void*) array, (size_t) count, sizeof ( struct lindexTitleRec ), (int (*) (const void*, const void*)) vmrrLindexSort ) ;

    for ( i = 0 ; i < count ; i++ )
    {
      if ( yropt == YR && mrropt == NONE )
      {
        displayTitleYear ( array [ i ] . title, array [ i ] . year ) ;
        (void) printf ( "\n" ) ;
      }
      else if ( yropt == YR )
      {
        if ( array [ i ] . mrr . votes == 0 )
        {
          if ( !constraints -> mrrMatch )
          {
	    (void) printf ( "                                " ) ;
            displayTitleYear ( array [ i ] . title, array [ i ] . year ) ;
            (void) printf ( "\n" ) ;
          }
        }
        else
        {
          (void) printf ( "      %s%8d   %2.1f  ", array [ i ] . mrr . distribution, array [ i ] . mrr . votes, array [ i ] . mrr . rating ) ;
          displayTitleYear ( array [ i ] . title, array [ i ] . year ) ;
          (void) printf ( "\n" ) ;
	}
      }
      else
      {
        if ( array [ i ] . mrr . votes == 0 )
        {
          if ( !constraints -> mrrMatch )
	    (void) printf ( "                                %s\n", array [ i ] . title ) ;
        }
        else
          (void) printf ( "      %s%8d   %2.1f  %s\n", array [ i ] . mrr . distribution, array [ i ] . mrr . votes, array [ i ] . mrr . rating, array [ i ] . title ) ;
      }
      free ( (void*) array [ i ] . title ) ;
    }
  }
       
  (void) fclose ( dbFp ) ;
  (void) fclose ( indexFp ) ;
  (void) fclose ( keyFp ) ;
  if ( mrropt != NONE )
    (void) fclose ( mrrFp ) ;
  for ( i = 0 ; i < NO_OF_TITLE_INFO_LISTS ; i++ )
  {
    if ( constraints -> titleInfoString [ i ] [ 0 ] )
    {
      (void) fclose ( titleInfoDb [ i ] ) ;
      (void) fclose ( titleInfoIdx [ i ] ) ;
    }
  }
}


int main ( int argc, char **argv )
{
  int err = FALSE ;
  int i, j ;
  int okopt ;
  int dbcount = 0 ;
  int yearOption = NONE ;
  int mrrOpt = NONE ;
  int nmopt = FALSE ;
  int ttlopt = FALSE ;
  struct lindexRec recs [ NO_OF_FILMOGRAPHY_LISTS ] ;
  struct searchConstraints constraints ;

  logProgram ( argv [ 0 ] ) ;
  initialiseConstraints ( &constraints ) ;

  if ( argc == 1 )
    err = TRUE ;

  for ( i = 1; i < argc; i++ )
    if ( strcmp ( "-i", argv[i] ) == 0 )
      constraints . caseSen = TRUE ;
    else if ( strcmp ( "-m", argv[i] ) == 0 )
      constraints . moviesOnly = TRUE ;
    else if ( strcmp ( "-s", argv[i] ) == 0 )
    {
      if ( ++i < argc )
	(void) strcpy ( constraints . subString, argv[i] ) ;
      else
	err = TRUE ;
    }
    else if ( strcmp ( "-mrr", argv[i] ) == 0 )
    {
      mrrOpt = MRR ;
      constraints . moviesOnly = TRUE ;
    }
    else if ( strcmp ( "-smrr", argv[i] ) == 0 )
    {
      mrrOpt = SMRR ;
      constraints . moviesOnly = TRUE ;
    }
    else if ( strcmp ( "-vmrr", argv[i] ) == 0 )
    {
      mrrOpt = VMRR ;
      constraints . moviesOnly = TRUE ;
    }
    else if ( strcmp ( "-match", argv[i] ) == 0 )
      constraints . mrrMatch = TRUE ;
    else if ( strcmp ( "-yr", argv[i] ) == 0 )
      yearOption = YR ;
    else if ( strcmp ( "-yru", argv[i] ) == 0 )
      yearOption = YRU ;
    else if ( strcmp ( "-title", argv[i] ) == 0 )
      ttlopt = TRUE ;
    else if ( strcmp ( "-yrto", argv[i] ) == 0 )
    {
      if ( ++i < argc )
        constraints . yearTo = atoi ( argv[i] ) ;
      else
        err = TRUE ;
    }
    else if ( strcmp ( "-yrfrom", argv[i] ) == 0 )
    {
      if ( ++i < argc )
        constraints . yearFrom = atoi ( argv[i] ) ;
      else
        err = TRUE ;
    }
    else if ( strcmp ( "-yreq", argv[i] ) == 0 )
    {
      if ( ++i < argc )
      {
        constraints . yearFrom = atoi ( argv[i] ) ;
        constraints . yearTo = constraints . yearFrom ;
      }
      else
        err = TRUE ;
    }
    else if ( strcmp ( "-vmax", argv[i] ) == 0 )
    {
      if ( mrrOpt == NONE )
        mrrOpt = MRR ;
      if ( ++i < argc )
        constraints . voteMax = atoi ( argv[i] ) ;
      else
        err = TRUE ;
    }
    else if ( strcmp ( "-vmin", argv[i] ) == 0 )
    {
      if ( mrrOpt == NONE )
        mrrOpt = MRR ;
      if ( ++i < argc )
        constraints . voteMin = atoi ( argv[i] ) ;
      else
        err = TRUE ;
      constraints . mrrMatch = TRUE ;
    }
    else if ( strcmp ( "-veq", argv[i] ) == 0 )
    {
      if ( mrrOpt == NONE )
        mrrOpt = MRR ;
      if ( ++i < argc )
      {
        constraints . voteMin = atoi ( argv[i] ) ;
        constraints . voteMax = constraints . voteMin ;
      }
      else
        err = TRUE ;
    }
    else
    {
      okopt = FALSE ;
      for ( j = 0 ; j < NO_OF_FILMOGRAPHY_LISTS ; j++ )
	if ( strcmp ( filmographyDefs [ j ] . option, argv [ i ] ) == 0 )
        {
	  okopt = TRUE ;
	  recs [ dbcount ] . fp = NULL ;
	  recs [ dbcount ] . listId = filmographyDefs [ j ] . listid ;
	  recs [ dbcount ] . name [ 0 ] = '\0' ;
	  dbcount++ ;
	  nmopt = TRUE ;
	}
      if ( ! okopt )
      {
        okopt = FALSE ;
        for ( j = 0 ; j < NO_OF_TITLE_INFO_LISTS ; j++ )
          if ( strcmp ( titleInfoDefs [ j ] . option, argv [ i ] ) == 0 )
          {
            okopt = TRUE ;
            if ( ++i < argc )
	      (void) strcpy ( constraints . titleInfoString [ j ], argv[i] ) ;
            else
	      err = TRUE ;
          }
      }
      if ( ! okopt )
      {
         (void) fprintf ( stderr, "lindex: unrecognised option %s\n", argv[i] ) ;
         err = TRUE ;
      }
    }

  if ( !err && !nmopt && !ttlopt )
    for ( j = 0 ; j < NO_OF_FILMOGRAPHY_LISTS ; j++ )
    {
       recs [ dbcount ] . fp = NULL ;
       recs [ dbcount ] . listId = filmographyDefs [ j ] . listid ;
       recs [ dbcount ] . name [ 0 ] = '\0' ;
       dbcount++ ;
       nmopt = TRUE ;
    }

  if ( nmopt && ttlopt )
  {
    (void) fprintf ( stderr, "lindex: can't mix title and name indexing in same command\n" ) ;
    err = TRUE ;
  }

  if ( !err && !nmopt && !ttlopt )
  {
    (void) fprintf ( stderr, "lindex: must specify at least one title or name indexing option\n" ) ;
    err = TRUE ;
  }


  if ( err )
   moviedbUsage ( LINDEX_USAGE1, LINDEX_USAGE2, LINDEX_USAGE3, LINDEX_USAGE4, LINDEX_USAGE5, LINDEX_USAGE6 ) ;

  if ( nmopt )
    lindexFilmographyDatabases ( dbcount, recs, &constraints ) ;
  else
    lindexTitles ( yearOption, mrrOpt, &constraints ) ; 
 
  return ( 0 ) ;
}
